contracts/src/finance/vesting_wallet.rs

```diff
@@ -0,0 +1,621 @@
+    /// implementation is a linear vesting curve.
+    ///
+    /// # Arguments
+    ///
+    /// * `&mut self` - Write access to the contract's state.
+    /// * `token` - Address of the token being released.
+    /// * `timestamp` - Point in time for which to check the vested amount.
+    ///
+    /// # Errors
+    ///
+    /// If the `token` address is not a contract, then the error
+    /// [`Error::InvalidToken`] is returned.
+    ///
+    /// # Panics
+    ///
+    /// If total allocation exceeds `U256::MAX`.
+    /// If scaled total allocation (mid calculation) exceeds `U256::MAX`.
+    #[selector(name = "vestedAmount")]
+    fn vested_amount_erc20(
+        &mut self,
+        token: Address,
+        timestamp: u64,
+    ) -> Result<U256, Self::Error>;
+}
+
+#[public]
+impl IVestingWallet for VestingWallet {
+    type Error = Error;
+
+    fn owner(&self) -> Address {
+        self.ownable.owner()
+    }
+
+    fn transfer_ownership(
+        &mut self,
+        new_owner: Address,
+    ) -> Result<(), Self::Error> {
+        Ok(self.ownable.transfer_ownership(new_owner)?)
+    }
+
+    fn renounce_ownership(&mut self) -> Result<(), Self::Error> {
+        Ok(self.ownable.renounce_ownership()?)
+    }
+
+    #[payable]
+    fn receive_ether(&self) {}
+
+    fn start(&self) -> U256 {
+        U256::from(self._start.get())
+    }
+
+    fn duration(&self) -> U256 {
+        U256::from(self._duration.get())
+    }
+
+    fn end(&self) -> U256 {
+        // SAFETY: both `start` and `duration` are stored as u64,
+        // so they cannot exceed `U256::MAX`
+        self.start() + self.duration()
+    }
+
+    #[selector(name = "released")]
+    fn released_eth(&self) -> U256 {
+        self._released.get()
+    }
+
+    #[selector(name = "released")]
+    fn released_erc20(&self, token: Address) -> U256 {
+        self._erc20_released.get(token)
+    }
+
+    #[selector(name = "releasable")]
+    fn releasable_eth(&self) -> U256 {
+        // SAFETY: total vested amount is by definition greater than or equal to
+        // the released amount.
+        self.vested_amount_eth(block::timestamp()) - self.released_eth()
+    }
+
+    #[selector(name = "releasable")]
+    fn releasable_erc20(
+        &mut self,
+        token: Address,
+    ) -> Result<U256, Self::Error> {
+        let vested = self.vested_amount_erc20(token, block::timestamp())?;
+        // SAFETY: total vested amount is by definition greater than or equal to
+        // the released amount.
+        Ok(vested - self.released_erc20(token))
+    }
+
+    #[selector(name = "release")]
+    fn release_eth(&mut self) -> Result<(), Self::Error> {
+        let amount = self.releasable_eth();
+        // SAFETY: total vested amount is by definition smaller than or equal to
+        // the total allocation, which is already verified to be smaller than
+        // `U256::MAX` in [`Self::releasable_eth`].
+        self._released.set(self._released.get() + amount);
```

Comment:
```suggestion
        let released = self._released.get().checked_add(amount).expect("should not fail");
        self._released.set(released);
```

Reply:
This should never panic, as `amount` and `_released` are equal to vested amount at this point in time, so if this were to ever panic, it would have happened in the [`releasable_eth` call](https://github.com/OpenZeppelin/rust-contracts-stylus/pull/402/files/ba278154b86aba20ae81f076f0741a7514f6de3e#diff-4083d07e30a6c3e16c834ce38c8c2236c57a943e7a89f8e4212209eb4a1bcd96R406).

In general, I'd be more in favor of a tiny perf. impact of using `checked_X` instead of relying on `SAFETY` comments (the code can change, and the comment becomes wrong, even harmful), but interested in your thoughts?

Reply:
I am more into `checked_X` on situations like here. Code may change in the future and someone may forget to apply checks here.